File: C++/FretBoard.cpp
#include "FretBoard.hpp"
#include "Note.hpp"
#include "utility_funcs.hpp"
#include "constants.hpp"
#include <algorithm>
#include <vector>
#include <cstddef>
#include <iostream>
#include <sstream>
#include <cmath>
#include <string>

using std::find_if; using std::vector;
using std::size_t; using std::ostream;
using std::sort; using std::ostringstream;
using std::log10; using std::floor;
using std::string; using std::endl;

FretBoard::FretBoard(const std::vector<Note>& notes) //constructor taking vector of Notes
  : pImpl(new FbImpl)
{
  if(!notes.empty())
    {
      for(vector<Note>::size_type i = 0; i != notes.size(); ++i)
	add(notes[i]);
    }
}

FretBoard& FretBoard::operator=(FretBoard rhs) //copy-assignment operator
{
  using std::swap;

  swap(rhs.pImpl, pImpl);
  return *this;
}

void FretBoard::add(const Note& n) //adds a new note
{
  superimpose(n, pImpl->data, true);

  vector<Note>::const_iterator i = pImpl->notes.begin();
  while(i != pImpl->notes.end() && i->name() != n.name() && i->alias() != n.name()) 
    ++i;

  if(i == pImpl->notes.end()) //then note n doesn't exist in Scale
    pImpl->notes.push_back(n);
}

void FretBoard::remove(const Note& n) //removes a note
{
  superimpose(n, pImpl->data, false);

  vector<Note>::iterator i = pImpl->notes.begin();
  while(i != pImpl->notes.end() && i->name() != n.name())
    ++i;
  
  if(i != pImpl->notes.end())
    pImpl->notes.erase(i);
}

void FretBoard::clear() //clears all existing notes
{
  pImpl.reset(new FbImpl);
  pImpl->notes.clear();
}

size_t FretBoard::num_notes() const
{
  return pImpl->notes.size();
}

void FretBoard::display_notes(ostream& os) const
{
  vector<string> temp;
  for(vector<Note>::size_type i = 0; i != pImpl->notes.size(); ++i)
    temp.push_back(pImpl->notes[i].name());
  
  sort(temp.begin(), temp.end());

  for(vector<string>::size_type i = 0; i != temp.size(); ++i)
    os << temp[i] << " ";
}

vector<string> FretBoard::make_string(size_t str_num, const string& s, size_t num_frets) const
{
  static const string emptyFret = "-----";
  static const string nonEmptyFret = "--\\--";
  
  vector<string> ret(1, s);
  for(vector<int>::size_type i = 0; i != num_frets; ++i)
    {
      if(pImpl->data[str_num][i] == 1)
	ret.push_back(nonEmptyFret);
      else
	ret.push_back(emptyFret);
    }
  return ret;
}

void FretBoard::display_grid(ostream& os, size_t num_frets) const
{
  static const string emptyDivider = "|    ";
  const vector<string> stringDivider(num_frets + 1, emptyDivider);

  vector<string> Estring = make_string(0, "E  -", num_frets);   //construct guitar strings
  vector<string> Astring = make_string(1, "A  -", num_frets);
  vector<string> Dstring = make_string(2, "D  -", num_frets);
  vector<string> Gstring = make_string(3, "G  -", num_frets);
  vector<string> Bstring = make_string(4, "B  -", num_frets);
  vector<string> estring = make_string(5, "e  -", num_frets);

  vector<string> newStringDivider = stringDivider;
  newStringDivider.insert(newStringDivider.begin(), string("   "));

  ostringstream oss; //construct fret number panel   
  oss << "      " << 1 << "  ";
  if(num_frets > 1)
    {
      for(size_t i = 2; i != num_frets + 1; ++i)
	oss << string(2 - (int)floor(log10(i)), ' ') << i << "  ";
    }

  vector<string> fretNumbers;
  fretNumbers.push_back(oss.str());

  vector<vector<string> > ret;   //construct grid to be displayed
  ret.push_back(fretNumbers);
  ret.push_back(Estring);
  ret.push_back(newStringDivider);
  ret.push_back(Astring);
  ret.push_back(newStringDivider);    
  ret.push_back(Dstring);
  ret.push_back(newStringDivider);
  ret.push_back(Gstring);
  ret.push_back(newStringDivider);
  ret.push_back(Bstring);
  ret.push_back(newStringDivider);
  ret.push_back(estring);

  for(vector<vector<string> >::size_type i = 0; i != ret.size(); ++i)   //output grid
    {
      for(vector<string>::size_type j = 0; j != ret[i].size(); ++j)
	os << ret[i][j];
      os << endl;
    }
} 

File: Java/Scale/FretBoard.java
package Scale;
import Scale.Note.*;
import java.util.*;
import Constants.*;

class FretBoard implements FretBoardInt
{
    protected FretBoard()
    {
	Impl = new FbImpl();
    }

    protected FretBoard(Note[] notes)
    {
	Impl = new FbImpl();
	if(notes.length != 0)
	    {
		for(int i = 0; i != notes.length; ++i)
		    add(notes[i]);
	    }
    }

    protected FretBoard(FretBoard Fb)
    {
	Impl = new FbImpl(Fb.Impl);
    }

    public void add(Note n)
    {
	Superimpose.superimpose(n, Impl.data, true);
	
	int i = 0;
	while(i != Impl.notes.length && Impl.notes[i].name().compareTo(n.name()) != 0)
	    ++i;

	if(i == Impl.notes.length)
	    {
		Note[] tmp = new Note[Impl.notes.length + 1];
		for(int j = 0; j != Impl.notes.length; ++j)
		    tmp[j] = Impl.notes[j];
		tmp[Impl.notes.length] = n;
		Impl.notes = tmp;
	    }
    }

    public void remove(Note n)
    {
	Superimpose.superimpose(n, Impl.data, false);

	int i = 0;
	while(i != Impl.notes.length && Impl.notes[i].name().compareTo(n.name()) != 0)
            ++i;

	if(i != Impl.notes.length)
	    {
		Note[] tmp = new Note[Impl.notes.length - 1];
		for(int j = 0, k = 0; j != Impl.notes.length; ++j)
		    {
			if(j != i)
			    tmp[k++] = Impl.notes[j];
		    }
		Impl.notes = tmp;
	    }
    }

    public void clear()
    {
	Impl = new FbImpl();
    }

    public int num_notes()
    {
	return Impl.notes.length;
    }

    public void display_notes()
    {
	String temp[] = new String[Impl.notes.length];
	for(int i = 0; i != Impl.notes.length; ++i)
	    temp[i] = Impl.notes[i].name();

	Arrays.sort(temp, String.CASE_INSENSITIVE_ORDER);

	for(int i = 0; i != Impl.notes.length; ++i)
	    System.out.print(temp[i] + " ");
    }

    private String[] make_string(int str_num, String s, int num_frets)
    {
	String[] ret = new String[num_frets + 2];
	ret[0] = s;
	for(int i = 0; i != num_frets; ++i)
	    {
		if(Impl.data[str_num][i] == 1)
		    ret[i+1] = nonEmptyFret;
		else
		    ret[i+1] = emptyFret;
	    }
	ret[num_frets+1] = "-";
	return ret;
    }

    public void display_grid(int num_frets)
    {
	if(num_frets > 0)
	    {
		String[] newStringDivider = new String[num_frets + 2];
		newStringDivider[0] = "   ";
		for(int i = 1; i != newStringDivider.length; ++i)
		    newStringDivider[i] = emptyDivider;
		
		String[] Estring = make_string(0, "E  ", num_frets);
		String[] Astring = make_string(1, "A  ", num_frets);
		String[] Dstring = make_string(2, "D  ", num_frets);	
		String[] Gstring = make_string(3, "G  ", num_frets);
		String[] Bstring = make_string(4, "B  ", num_frets);
		String[] estring = make_string(5, "e  ", num_frets);
		
		String fretNumbers = new String("     " + 1 + "  ");
		if(num_frets > 1)
		    {
			for(int i = 2; i != num_frets + 1; ++i)
			    {
				int tmp = 3 - (int)Math.floor(Math.log10(i));
				for(int j = 0; j != tmp; ++j) 
				    fretNumbers += " ";
				fretNumbers += i + " ";
			    }
		    }
		
		System.out.println(fretNumbers);
		display_string(Estring);
		display_string(newStringDivider);
		display_string(Astring);
		display_string(newStringDivider);
		display_string(Dstring);
		display_string(newStringDivider);
		display_string(Gstring);
		display_string(newStringDivider);
		display_string(Bstring);
		display_string(newStringDivider);
		display_string(estring);
	    }
    }

    private void display_string(String[] s)
    {
	for(int i = 0; i != s.length; ++i)
	    System.out.print(s[i]);
	System.out.println();
    }

    private FbImpl Impl;
    private static final String emptyFret = "-----";
    private static final String nonEmptyFret = "--\\--";
    private static final String emptyDivider = "|    ";
}